### DISCUSSION TITLE
1. https://www.google.com에 접속할 때 일어나는 일에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#1] [General]
웹 통신의 큰 흐름에 대해 학습해 봅시다.

**Comment by eun-byeol** [comment]
### 🔑 키워드
`도메인 이름 IP 주소 변환`, `host 파일`, `DNS 캐시`, `DNS 서버`, `3-way handshake`, `SSL/TLS handshake`, `HTTP 요청 메시지`, `TCP 프로토콜`, `HTTP 응답 메시지`, `랜더링`

---

1. 브라우저는 사용자가 입력한 URL을 파싱합니다.
2. www.google.com 이라는 `도메인 이름을 IP 주소로 변환`하기 위해, 로컬 PC의 `host 파일`과 `DNS 캐시`를 확인합니다. 정보가 없다면, `DNS 서버`에 쿼리를 보내 IP 주소를 찾습니다.
3. TCP 연결을 위해 `3-way handshake`를 수행한 후, HTTPS 프로토콜의 보안을 위해 `SSL/TLS handshake`를 진행합니다.
4. 브라우저는 `TCP 프로토콜`을 통해 생성한 `HTTP 요청 메시지`를 구글 서버로 전송합니다.
5. 구글 서버는 요청을 처리하고 `HTTP 응답 메시지`를 반환합니다.
6. 브라우저는 수신한 응답 데이터를 파싱하여 DOM*을 생성하고, `랜더링`하여 사용자 컴퓨터 화면에 출력합니다.

![image](https://github.com/user-attachments/assets/f99d4855-c2a6-456b-b047-2aa6b85c5d81)

---
### 더 알아보기
백엔드에게 DOM 내용은 불필요하면 언급하지 않아도 좋을 것 같아요. 참고만 해주세요~
> DOM(Document Object Model : 문서 객체 모델)

HTML 문서를 기반으로 브라우저가 웹 페이지의 구조와 내용을 이해하고 `조작`할 수 있도록 만든 `구조화된 트리 모델`
개발자는 자바스크립트와 같은 스크립팅 언어를 사용해 DOM을 조작하거나 수정할 수 있다.
비유: HTML이 건물 설계도라면, DOM은 브라우저가 설계도를 읽어 만든 3D 건물의 모형. 개발자는 모형의 창문을 열거나 닫고, 문의 위치를 바꾸는 등 작업을 할 수 있다.
